04/r/particulate-matter.Rmd

```diff
@@ -0,0 +1,385 @@
+
+```{r informative-prior, comment=NA}
+prior_file <- "../../stan/particulate_matter_ols_prior.stan"
+cat(read_file(prior_file))
+```
+
+Compile this stan model.
+```{r stan-compile-priors}
+prior_model <- stan_model(prior_file)
+```
+
+Now let's sample from this model.
+```{r stan-sample-priors}
+prior_samples <- sampling(prior_model, data = prior_data, iter = 300, chains = 1)
+```
+
+```{r prior-extract-samples}
+log_pm_prior <- prior_samples %>% 
+    spread_draws(log_pm_prior[city_id]) %>% 
+    ungroup()
+
+log_pm_prior
+```
+
+Let's take a look at 9 random draws from our estimates. We'll also join in the original data (by `city_id`) so we can compare against.
+```{r prior-log-pm-comparison}
+log_pm_comparison <- log_pm_prior %>% 
+    filter(.draw %in% sample(max(.draw), size = 9L)) %>% 
+    left_join(data %>% select(city_id, log_sat, log_pm25), by = "city_id") %>% 
+    select(.draw, city_id, log_sat, observed = log_pm25, sampled = log_pm_prior)
+
+log_pm_comparison
+```
+
+Let's visualize these results to see if our prior distribution is producing feasible values for `log_pm25`.
+```{r plot-prior-log-pm-comparison}
+log_pm_comparison %>% 
+    gather("type", "log_pm_25", observed:sampled) %>% 
+    ggplot(aes(x = log_sat, y = log_pm_25, color = type)) +
+    geom_point(alpha = 0.3) + 
+    facet_wrap(~ .draw)
+```
+
+## 4. Baseline Model Inference
+Here's Gabry *et al's* code for the basic regression model:
+
+```{r regression-model, comment=NA}
+model_file <- "../../stan/particulate_matter_ols.stan"
+cat(read_file(model_file))
+```
+Compile this stan model.
+```{r stan-compile-regression}
+model <- stan_model(model_file)
+```
+
+Now let's sample from this model.
+```{r stan-sample-regression}
+model_data <- list(N = nrow(data), log_sat = data$log_sat, log_pm = data$log_pm25)
+model_samples <- sampling(model, data = model_data, iter = 1000, chains = 4)
+```
+
+And extract the draws for $\beta_0$, $\beta_1$, and $\sigma$.
+```{r model-extract-samples}
+coef_samples <- model_samples %>% 
+    spread_draws(beta0, beta1, sigma) %>% 
+    ungroup()
+
+coef_samples
+```
+
+### 4.1 Trace Plots
+Gabry's paper has some slick visualizations for investigating your MCMC run, that we'll talk about in the future when we discuss Hamiltonian Monte Carlo. For now, just do some quick trace plots to ensure nothing terrible happened:
+
+```{r plot-trace}
+coef_samples %>% 
+    gather("parameter", "value", beta0:sigma) %>% 
+    ggplot(aes(x = .draw, y = value)) +
+    geom_line() + 
+    facet_wrap(~parameter, ncol = 1, scales = "free_y")
+```
+
+Looks like we're mixing OK.
+
+Not surprisingly, the slope and intercept show some negative correlation:
+
+```{r plot-coef-variation}
+ggplot(coef_samples, aes(x = beta0, y = beta1)) + geom_hex() + scale_fill_viridis_c()
```

Comment:
This requires the hexbin library - need to make sure folks have that installed ahead of time